src/imp.rs

```diff
@@ -526,6 +509,45 @@ where
     }
 }
 
+/// Pick the notes to be used in a transaction from a vector of notes.
+///
+/// The notes are picked in a way to maximize the number of notes used, while
+/// best-effort minimizing the value employed. To do this we sort the notes in
+/// ascending value order, and go through each contiguous slice of the largest
+/// possible size until we find the first slice whose sum is larger or equal to
+/// the given value. If such a slice is not found, None is returned.
+///
+/// Note: it is presupposed that the input notes contain enough balance to cover
+/// the given `value`.
+fn pick_notes(
+    value: u64,
+    mut notes_and_values: Vec<(Note, u64, JubJubScalar)>,
+) -> Option<Vec<(Note, u64, JubJubScalar)>> {
+    let len = notes_and_values.len();
+
+    if len <= MAX_INPUT_NOTES {
+        return Some(notes_and_values);
+    }
+
+    notes_and_values.sort_by(|(_, aval, _), (_, bval, _)| aval.cmp(bval));
+
+    let mut picked = Vec::with_capacity(MAX_INPUT_NOTES);
+
+    for i in 0..len - MAX_INPUT_NOTES {
+        if notes_and_values[i..i + MAX_INPUT_NOTES]
+            .iter()
+            .map(|(_, val, _)| val)
+            .sum::<u64>()
+            >= value
+        {
+            picked.extend(notes_and_values.drain(i..i + MAX_INPUT_NOTES));
```

Comment:
Even with this approach I don't think `picked` is actually ever needed. You could just have:
```rust
return Some(
    notes_and_values.drain(i..i + MAX_INPUT_NOTES).collect(),
);
```
(or you could have a local `picked` variable, but not need to preallocate the capacity at this point).